sdk/src/main/java/io/radar/sdk/util/RadarLogBuffer.kt

```diff
@@ -0,0 +1,245 @@
+package io.radar.sdk.util
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import io.radar.sdk.Radar
+import io.radar.sdk.model.RadarLog
+import org.json.JSONObject
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.util.*
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ * Wraps the filesystem to support log caching
+ */
+internal class RadarLogBuffer(
+    /**
+     * Application Context
+     */
+    context: Context,
+    /**
+     * Executor for writing files. This is used to ensure that log entries are written to the file in the order that
+     * they are received.
+     */
+    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
+) {
+
+    private companion object {
+
+        /**
+         * Name of the directory in which the logs are stored
+         */
+        const val DIRECTORY = "RadarLogs"
+
+        /**
+         * Maximum number of logs in the buffer.
+         */
+        const val LOG_BUFFER_SIZE_LIMIT = 500
+    }
+
+    /**
+     * The directory where the log files are stored
+     */
+    private val directory = context.getDir(DIRECTORY, Context.MODE_PRIVATE)
+
+    /**
+     * Lock for handling changes to [currentFile]
+     */
+    private val fileLock = ReentrantLock()
+
+    /**
+     * Lock for handling changes to the files within [DIRECTORY]
+     */
+    private val directoryLock = ReentrantLock()
+
+    /**
+     * Contains the name of the current log file. Modifications to this value should occur within a [fileOp]
+     * operation.
+     */
+    private var currentFile: String
+
+    /**
+     * Number of log entries in the current log file. Modifications to this value should occur within a [fileOp]
+     * operation.
+     */
+    var size: Int
+        private set
+
+    init {
+        //First find current file
+        val files = directory.listFiles()
+        if (files?.isNotEmpty() == true) {
+            //This gets the latest file that was in use if it can, otherwise it returns the first one.
+            currentFile = files.maxByOrNull { getLastModified(it) }!!.name
+            //This is initialized to the number of lines in the current file
+            size = FileInputStream(File(directory, currentFile)).bufferedReader().readLines().size
+            if (files.size > 2) {
```

Comment:
given there's no lock on this + purgeInactiveFiles already checks for files.size > 2, do we need this check?

Reply:
If we purge every time this is reached, a new file will get created unnecessarily (since `currentFile` is assigned to a new file in the `purgeInactiveFiles` method). I have some unit tests in place for this scenario, and verified that this change causes that side-effect. By doing the check in the constructor, it allows the existing file to be used for writing additional logs. LMK if you still feel like this should be changed. As for not using the lock here - since this occurs in the constructor, other methods can't be called on the object at this time, so the lock seemed unnecessary.